File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.cpp
class Solution {
public:
    int height;
    int width;
    int move[4][2] = {{0, 1},{1, 0},{0, -1},{-1, 0}};
    
    void helper(vector<vector<int>>& nums, vector<vector<bool>>& v, int x, int y) {
        
        if (v[x][y]) return;
        
        v[x][y] = true;
        
        for (int i = 0; i < 4; i++) {
            int p = x + move[i][0];
            int q = y + move[i][1];
            
            if (p < 0 || q < 0 || p >= height || q >= width) continue;
            if (nums[p][q] >= nums[x][y])
                helper(nums, v, p, q);
        }
        
    } 
    
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        
        vector<vector<bool>> pacific;
        vector<vector<bool>> atlan;
        vector<vector<int>> ans;
        height = heights.size();
        width = heights[0].size();
        
        pacific.resize(height, vector<bool>(width, false));
        atlan.resize(height, vector<bool>(width, false));
        
        for (int i = 0; i < width; i++) {
            if (!pacific[0][i]) helper(heights, pacific, 0, i);
            if (!atlan[height - 1][i]) helper(heights, atlan, height - 1, i);
        }
        for (int i = 0; i < height; i++) {
            if (!pacific[i][0]) helper(heights, pacific, i, 0);
            if (!atlan[i][width - 1]) helper(heights, atlan, i, width - 1); 
        }
        
        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++)
                if (pacific[i][j] && atlan[i][j])
                    ans.push_back({i, j});
        
        return ans;
    }
};

File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, grid: List[List[int]]) -> List[List[int]]:
        
        answer = []
        height, width = len(grid), len(grid[0])
        movs = [ [1,0],[-1,0],[0,1],[0,-1] ]
        
        p = [[False for col in range(width)] for row in range(height)]
        c = [[False for col in range(width)] for row in range(height)]
        
        def solve(island: List[List[int]], x: int, y: int): # no return
            
            if island[x][y]: 
                return
            
            island[x][y] = True
            
            if p[x][y] and c[x][y]:
                answer.append([x, y])
            for mov in movs:
                nx, ny = x + mov[0], y + mov[1]
                
                if nx < 0 or ny < 0 or nx >= height or ny >= width or island[nx][ny]:
                    continue
                if grid[x][y] <= grid[nx][ny]:
                    solve(island, x + mov[0], y + mov[1])
        
        for i in range(height):
            solve(p, i, 0)
        for i in range(width):
            solve(p, 0, i)
        for i in range(height):
            solve(c, i, width -1)
        for i in range(width):
            solve(c, height - 1, i)
        
        return answer
    